feat: add glossary to packaging guide (#635)
pyOpenSci/python-package-guide: documentation/glossary.md

# Python packaging glossary

```{eval-rst}
:og:description: A glossary of common terms used across this Python package
  guide.
:og:title: Python packaging glossary
```

## Core packaging

```{glossary}

> __init__.py

Reviewer: ```suggestion
`__init__.py`
```

  A special Python file that marks a directory as a Python package.
  When Python sees this file, it knows the folder contains importable
  code. It can be empty or contain code that runs when the package is

>   imported.

Reviewer 2: ```suggestion
  code. It can either be empty or contain code that runs when the package
  is imported.
```

API token

>   A secret key used to authenticate with PyPI or Test PyPI when

Reviewer 2: ```suggestion
  A secret key used to authenticate with PyPI or TestPyPI when
```

  publishing a package. You generate one in your account settings and
  use it in place of a password. Treat it like a password and never

>   share it or commit it to version control.

Reviewer: ```suggestion
  use it in place of a password. **Treat it like a password and never
  share it or commit it to version control**.
```

Build backend
  The tool that does the actual work of building your package into
  distribution files. In this guide, the build backend is Hatchling.
  You specify it in your `pyproject.toml` file and it runs when you run

>   `hatch build`.

Reviewer 2: ```suggestion
  `hatch build` or `python -m build`
```

The reasoning for this point is considering that it is not standard practice to install the build backend (or the package that provides the build backend in this case, which is `hatch`) directly – the build frontend (PyPA/build here or `pip wheel` otherwise as well) manages this as part of a temporary build-time venv, etc.

Reviewer: I think having both is fine.

To use hatchling to build a package, either are valid commands. I don't think the extra detail beyond mentioning both is needed here.

>   `hatch build`.

Reviewer: ```suggestion
  The tool that does the actual work of building your package into
  distribution files. In this guide, the build backend is Hatchling.
  You specify it in your `pyproject.toml` file:
  
  [build-system]
  requires = ["hatchling >= 1.26"]
  build-backend = "hatchling.build"

  You execute a build by running `hatch build`. Alternatively, you can run  `python -m build`.
  [Reference: official Python Packaging documentation](https://packaging.python.org/en/latest/tutorials/packaging-projects/#choosing-a-build-backend)
```

I couldn't make the toml block code style in a suggestion.

Distribution files
  The files you upload to PyPI so others can install your package.
  There are two common types: a wheel (`.whl`) and a source
  distribution (`.tar.gz`). See also `Wheel (.whl)` and `Source

>   distribution (sdist)`.

Reviewer 2: I think we should also mention somewhere that "a wheel" is the same as "binary distribution".

Module
  A single Python file (`.py`) containing code such as functions,
  classes, or variables that can be imported. A package is made up of
  one or more modules.

> pyproject.toml

Reviewer: ```suggestion
`pyproject.toml`
```

  The configuration file at the root of your Python package. Written in
  TOML format, it stores metadata such as name, version, authors, and
  license. It can also configure tools such as Hatch, uv, and pytest.

>   See also [Make your Python package PyPI ready](../tutorials/pyproject-toml).

Reviewer 2: ```suggestion
pyproject.toml
  The configuration file at the root of your Python package. Written in
  TOML format, it stores metadata such as name, version, authors, and
  license. It can also configure tools such as Hatch, uv, ruff, and pytest.
  See also [Make your Python package PyPI ready](../tutorials/pyproject-toml).
```

since we mention `ruff` below.

Also, maybe this should arrive after TOML?

Python package
  A directory of Python code structured so it can be installed,
  imported, and shared with others. A package includes at least an

>   `__init__.py` file and a `pyproject.toml` file.

Reviewer 2: It is not strictly required for a Python package to contain `__init__.py` files, so even for users browsing through this guide and building their first package, I think we should explain, or at the very least, make a brief mention of the difference, as this is a _white lie_ of sorts :) Python packages without `__init__.py` files also exist, and are called implicit namespace packages as defined by PEP 420 (see [PEP](https://peps.python.org/pep-0420/), and [core specification docs](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages))

>   `__init__.py` file and a `pyproject.toml` file.

Reviewer: ```suggestion
Python package
  A directory of Python code structured so it can be installed,
  imported, and shared with others. A package includes at least an
  `__init__.py` file and a `pyproject.toml` file. This is sometimes
  referred to as a **regular package**.
  
  Info: You may hear the term **namespaced package** which is not really
  a package at all but a container of subpackages. This is out of scope
  for this guide. If interested, consult the [Python documentation](https://docs.python.org/3/glossary.html#term-namespace-package).
```

PyPI / Test PyPI
  PyPI (the Python Package Index) is the official repository where

>   Python packages are published and installed from. Test PyPI is a

Reviewer 2: ```suggestion
PyPI / TestPyPI
  PyPI (the Python Package Index) is the official repository where
  Python packages are published and installed from. TestPyPI is a
```

  separate practice environment used for learning and testing publish

>   workflows. See [pypi.org](https://pypi.org) and

Reviewer 2: ```suggestion
  separate practice environment used for learning and testing publishing
  workflows. See [pypi.org](https://pypi.org) and
```

  [test.pypi.org](https://test.pypi.org).
  See also [Publish your Python package to PyPI](../tutorials/publish-pypi).

Source distribution (sdist)
  One of the two distribution file types for a Python package. The
  sdist (`.tar.gz`) contains source code and project files. When someone
  installs from an sdist, tools build the package locally first.
  See also [Publish your Python package to PyPI](../tutorials/publish-pypi).

TOML
  Tom's Obvious Minimal Language, a simple format for configuration
  files. TOML organizes data into tables such as `[project]` or
  `[tool.hatch]` and arrays. `pyproject.toml` uses TOML.

Trusted publishing
  A secure way to publish to PyPI using GitHub Actions instead of an
  API token. Rather than storing a secret token, you configure PyPI to
  trust your repository directly.
  See also [Setup Trusted Publishing for secure and automated publishing via GitHub Actions](../tutorials/trusted-publishing).

> Wheel (.whl)

Reviewer 2: ```suggestion
Wheel (.whl) (binary distribution)
```

as noted above!

  One of the two distribution file types for a Python package. A wheel
  is a pre-built format that installs directly without a build step.
  For many pure Python packages, one wheel can work across platforms.

>   See also [Publish your Python package to PyPI](../tutorials/publish-pypi).

Reviewer 2: ```suggestion
  One of the two distribution file types for a Python package. A wheel
  is a pre-built binary format (`.whl`, a ZIP file) that installs directly
  without a build step. For many pure Python packages, one wheel can
  work across platforms.
  See also [Publish your Python package to PyPI](../tutorials/publish-pypi).
```

for consistency: since we mentioned the `.tar.gz` format above for sdists, we can note this for wheels here. Please feel free to adjust this.

```

## Tools

```{glossary}
copier
  A command-line tool for creating new projects from templates. In this
  guide, you can use copier with the pyOpenSci package template to set
  up structure, configuration, and tooling quickly. See
  [copier.readthedocs.io](https://copier.readthedocs.io).

coverage.py
  A tool that measures how much of your code is exercised by tests,
  often as a percentage. It shows which lines and branches are covered.
  See [coverage.readthedocs.io](https://coverage.readthedocs.io).

Hatch
  A modern Python packaging and project management tool. In this guide,
  Hatch is used to build packages, manage environments, run scripts, and
  publish. Configuration lives in `pyproject.toml`. See
  [hatch.pypa.io](https://hatch.pypa.io).
  See also [Get to know Hatch](../tutorials/get-to-know-hatch).

> 
> Hatchling
>   The build backend used by Hatch. When you run `hatch build`,
>   Hatchling reads `pyproject.toml` and creates wheel and sdist files.
>   See [hatch.pypa.io/latest/backend](https://hatch.pypa.io/latest/backend/).

Reviewer 2: ```suggestion
Hatchling
  The build backend used by Hatch. When you run `python -m build` or
  `hatch build`, Hatchling reads `pyproject.toml` and creates sdist
  and wheel files.
  See [hatch.pypa.io/latest/backend](https://hatch.pypa.io/latest/backend/).
```

nit: I reordered "wheel" ↔️ "sdist" because the sdist is built first and then the wheel, from the built sdist.

(unless if you pass `--sdist` and `--wheel` separately, in which case they are built from the source tree, but the readers don't need to know that just yet :D)

pip
  Python's default package installer. You can use it to install packages
  from PyPI into an environment with commands such as
  `pip install package-name`. See [pip.pypa.io](https://pip.pypa.io).

pytest
  The standard Python testing framework used to discover and run tests.
  In this guide, pytest often runs through Hatch scripts. See

>   [docs.pytest.org](https://docs.pytest.org).

Reviewer: ```suggestion
  A widely used Python testing framework for discovering and runnning tests.
  In this guide, pytest often runs through Hatch scripts. See
  [docs.pytest.org](https://docs.pytest.org).
```

> Ruff
>   A fast Python linter and formatter. It checks style and can
>   automatically fix many issues. See
>   [docs.astral.sh/ruff](https://docs.astral.sh/ruff).

Reviewer 2: ```suggestion
Ruff
  A fast Python linter and formatter. It checks style and can
  automatically fix many styling issues. See
  [docs.astral.sh/ruff](https://docs.astral.sh/ruff).
```

Sphinx
  A documentation generator for Python projects. Sphinx reads docstrings
  and documentation files to build a docs site. See
  [sphinx-doc.org](https://www.sphinx-doc.org).

> Twine
>   A tool for securely uploading distribution files to PyPI or Test PyPI.
>   See [twine.readthedocs.io](https://twine.readthedocs.io).

Reviewer 2: ```suggestion
Twine
  A tool for securely uploading distribution files to PyPI or TestPyPI.
  See [twine.readthedocs.io](https://twine.readthedocs.io).
```

uv
  A fast Python package and environment manager. In this guide, you can
  use uv to manage dependencies and run commands in project environments.
  See [docs.astral.sh/uv](https://docs.astral.sh/uv).
```

## Hatch-specific concepts

```{glossary}
Hatch environment
  An isolated Python environment managed by Hatch. You can define
  multiple environments in `pyproject.toml` for testing, docs, builds,
  and style checks, each with its own dependencies and scripts.

Script (Hatch)
  A named command defined inside a Hatch environment in
  `pyproject.toml`. Scripts provide shortcuts such as
  `hatch run build:check` and `hatch run test:run`.

Task runner
  A tool that automates repetitive development workflows. Hatch can
  function as a task runner by letting you define scripts that run in
  specific environments.
```

## Development concepts

```{glossary}
Code coverage
  A measure of how much source code executes during tests, usually as a
  percentage. High coverage does not guarantee no bugs, but low coverage

>   can indicate untested areas.

Reviewer 2: would it make sense to make a note about "test coverage" too? See https://www.honeybadger.io/blog/code-test-coverage-python/, as they are related but different concepts that I believe will be useful for the reader.

Reviewer: @agriyakhetarpal If we were doing a course on testing, I would recommend explaining test coverage. While they are different topics, though tangential, I would stick with the more common code coverage here to reduce the learner's cognitive load. :smile:

Dependencies
  Other Python packages needed for your package to work. Common classes
  include required dependencies, optional dependencies, and development
  dependencies.

Docstring
  A string at the top of a function, class, or module that describes
  behavior, inputs, and outputs. Docstrings can be used by tools such as
  Sphinx to generate API documentation.

End-to-end test
  A test that simulates a complete user workflow from start to finish.
  In scientific packages, tutorials executed during docs builds can
  serve as end-to-end tests.

Integration test
  A test that checks how multiple functions or components work together.
  Unlike a unit test, it verifies behavior across a broader workflow.

Linting
  Automatic checks for style issues, formatting problems, and potential
  errors in code.

Unit test
  A test that checks one function or method in isolation. Unit tests are
  fast and help pinpoint where failures occur.

Version specifier / lower bound
  A constraint on which dependency versions are accepted. For example,
  `numpy>=1.24` sets a lower bound so versions older than 1.24 are not
  used.
```

## Git / GitHub

```{glossary}

> GitHub Codespace

Reviewer: ```suggestion
git
  A tool for version control.

GitHub
  A service providing accounts and organizations to facilitate sharing repositories.

GitHub Codespace
```

---
:og:title: Python packaging glossary
:og:description: A glossary of common terms used across this Python package
  guide.
---

> 

Reviewer: ```suggestion

# Python packaging glossary

```

# Python packaging glossary

```{eval-rst}
:og:description: A glossary of common terms used across this Python package
  guide.
:og:title: Python packaging glossary

> ```

Reviewer: ```suggestion
---
:og:title: Python packaging glossary
:og:description: A glossary of common terms used across this Python package
  guide.
---
```

## Core packaging

```{glossary}
`__init__.py`
  A special Python file that marks a directory as a Python package.
  When Python sees this file, it knows the folder contains importable
  code. It can either be empty or contain code that runs when the package
  is imported.

API token
  A secret key used to authenticate with PyPI or TestPyPI when
  publishing a package. You generate one in your account settings and
  use it in place of a password. **Treat it like a password and never
  share it or commit it to version control**.

Build backend
  The tool that does the actual work of building your package into
  distribution files. In this guide, the build backend is Hatchling.
  You specify it in your `pyproject.toml` file:

     [build-system]
     requires = ["hatchling >= 1.26"]
     build-backend = "hatchling.build"

>   You execute a build by running `hatch build`. Alternatively, you can run  `python -m build`.

Reviewer: ```suggestion
  You specify it in your `pyproject.toml` file under `[build-system]`.

  You execute a build by running `hatch build`. Alternatively, you can run  `python -m build`.
```

> 
>   [build-system]
>   requires = ["hatchling >= 1.26"]
>   build-backend = "hatchling.build"
> 
>   You execute a build by running `hatch build`. Alternatively, you can run  `python -m build`.
>   [Reference: official Python Packaging documentation](https://packaging.python.org/en/latest/tutorials/packaging-projects/#choosing-a-build-backend)

Reviewer: ```suggestion
Build backend
  The tool that does the actual work of building your package into
  distribution files. In this guide, the build backend is Hatchling.
  You specify it in your `pyproject.toml` file:

     [build-system]
     requires = ["hatchling >= 1.26"]
     build-backend = "hatchling.build"

  You execute a build by running `hatch build`. Alternatively, you can run  `python -m build`.
  [Reference: official Python Packaging documentation](https://packaging.python.org/en/latest/tutorials/packaging-projects/#choosing-a-build-backend)
```

  See also [Publish your Python package to PyPI](../tutorials/publish-pypi).

Source distribution (sdist)
  One of the two distribution file types for a Python package. The
  sdist (`.tar.gz`) contains source code and project files. When someone
  installs from an sdist, tools build the package locally first.
  See also [Publish your Python package to PyPI](../tutorials/publish-pypi).

TOML
  Tom's Obvious Minimal Language, a simple format for configuration
  files. TOML organizes data into tables such as `[project]` or
  `[tool.hatch]` and arrays. `pyproject.toml` uses TOML.

Trusted publishing
  A secure way to publish to PyPI using GitHub Actions instead of an
  API token. Rather than storing a secret token, you configure PyPI to
  trust your repository directly.
  See also [Setup Trusted Publishing for secure and automated publishing via GitHub Actions](../tutorials/trusted-publishing).

Wheel (.whl) (binary distribution)
  One of the two distribution file types for a Python package. A wheel
  is a pre-built binary format (`.whl`, a ZIP file) that installs directly
  without a build step. For many pure Python packages, one wheel can
  work across platforms.

>   See also [Publish your Python package to PyPI](../tutorials/publish-pypi).

Reviewer: ```suggestion
Wheel (.whl)
  The binary distribution type for a Python package. A wheel
  is a pre-built binary format (`.whl`, a ZIP file) that installs directly
  without a build step. For many pure Python packages, one wheel can
  work across platforms.
  See also [Publish your Python package to PyPI](../tutorials/publish-pypi).
```

  guide, you can use copier with the pyOpenSci package template to set
  up structure, configuration, and tooling quickly. See
  [copier.readthedocs.io](https://copier.readthedocs.io).

coverage.py
  A tool that measures how much of your code is exercised by tests,
  often as a percentage. It shows which lines and branches are covered.
  See [coverage.readthedocs.io](https://coverage.readthedocs.io).

Hatch
  A modern Python packaging and project management tool. In this guide,
  Hatch is used to build packages, manage environments, run scripts, and
  publish. Configuration lives in `pyproject.toml`. See
  [hatch.pypa.io](https://hatch.pypa.io).
  See also [Get to know Hatch](../tutorials/get-to-know-hatch).

Hatchling
  The build backend used by Hatch. When you run `python -m build` or
  `hatch build`, Hatchling reads `pyproject.toml` and creates sdist
  and wheel files.
  See [hatch.pypa.io/latest/backend](https://hatch.pypa.io/latest/backend/).

pip
  Python's default package installer. You can use it to install packages
  from PyPI into an environment with commands such as
  `pip install package-name`. See [pip.pypa.io](https://pip.pypa.io).

pytest

>   A widely used Python testing framework for discovering and runnning tests.

Reviewer: ```suggestion
  A widely used Python testing framework for discovering and running tests.
```
